jina/peapods/networking.py

```diff
@@ -1,8 +1,341 @@
+            return self._send_message(msg, pooled_connection)
+        elif self._on_demand_connection:
+            # If the pool is disabled and an unknown connection is requested: create it
+            connection_pool = self._create_connection_pool(target_address)
+            return self._send_message(msg, connection_pool.get_connection())
+        else:
+            raise ValueError(f'Unknown address {target_address}')
+
+    def _create_connection_pool(self, target_address):
+        port = target_address[target_address.rfind(':') + 1 :]
+        connection_pool = ConnectionList(port=port)
+        connection_pool.add_connection(
+            target_address, self._create_connection(target=target_address)
+        )
+        self._connections[target_address] = connection_pool
+        return connection_pool
+
+    def start(self):
+        """
+        Starts the connection pool
+        """
+        pass
+
+    def close(self):
+        """
+        Closes the connection pool
+        """
+        self._connections.clear()
+
+    @abstractmethod
+    def _send_message(self, msg: Message, connection):
+        ...
+
+    @abstractmethod
+    def _create_connection(self, target):
+        ...
+
+
+class GrpcConnectionPool(ConnectionPool):
+    """
+    GrpcConnectionPool which uses gRPC as the communication mechanism
+    """
+
+    def _send_message(self, msg: Message, connection):
+        # this wraps the awaitable object from grpc as a coroutine so it can be used as a task
+        # the grpc call function is not a coroutine but some _AioCall
+        async def task_wrapper(new_message, stub):
+            await stub.Call(new_message)
+
+        return asyncio.create_task(task_wrapper(msg, connection))
+
+    def _create_connection(self, target):
+        self._logger.debug(f'create connection to {target}')
+        channel = grpc.aio.insecure_channel(
+            target,
+            options=[
+                ('grpc.max_send_message_length', -1),
+                ('grpc.max_receive_message_length', -1),
+            ],
+        )
+
+        return jina_pb2_grpc.JinaDataRequestRPCStub(channel)
+
+
+class K8sGrpcConnectionPool(GrpcConnectionPool):
+    """
+    Manages grpc connections to replicas in a K8s deployment.
+
+    :param namespace: K8s namespace to operate in
+    :param client: K8s client
+    :param logger: the logger to use
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        client,
+        logger: JinaLogger = None,
+    ):
+        super().__init__(logger=logger, on_demand_connection=False)
+
+        self._namespace = namespace
+        self._deployment_clusteraddresses = {}
+        self._k8s_client = client
+        self._k8s_event_queue = asyncio.Queue()
+        self.enabled = False
+
+        self._fetch_initial_state()
+
+        from kubernetes import watch
+
+        self._api_watch = watch.Watch()
+
+        self.update_thread = Thread(target=self.run, daemon=True)
```

Comment:
I panic when I see `threads` but I guess there is no better alternative

Reply:
yeah its annoying, but we need the `Thread` for the blocking call in the kubernetes API due to https://github.com/kubernetes-client/python/issues/323